File: e-goi-wrapper-java/src/com/egoi/api/wrapper/impl/xmlrpc/EgoiApiXmlRpcImpl.java
package com.egoi.api.wrapper.impl.xmlrpc;

import com.egoi.api.wrapper.api.EgoiMap;
import com.egoi.api.wrapper.api.EgoiMapList;
import com.egoi.api.wrapper.api.exceptions.EgoiException;

public class EgoiApiXmlRpcImpl extends AbstractXmlRpcEgoiApi {

    private static final String ServiceUrl = "http://api.e-goi.com/v2/xmlrpc.php";

    public EgoiApiXmlRpcImpl() throws EgoiException {
        super(ServiceUrl);
    }

    @Override
    public EgoiMap addExtraField(EgoiMap arguments) throws EgoiException {
        return processRequest("addExtraField", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap addSubscriber(EgoiMap arguments) throws EgoiException {
        return processRequest("addSubscriber", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap addSubscriberBulk(EgoiMap arguments) throws EgoiException {
        return processRequest("addSubscriberBulk", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap getUserData(EgoiMap arguments) throws EgoiException {
        return processRequest("getUserData", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMapList getLists(EgoiMap arguments) throws EgoiException {
        return processRequest("getLists", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMap checkLogin(EgoiMap arguments) throws EgoiException {
        return processRequest("checkLogin", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap createCampaignEmail(EgoiMap arguments) throws EgoiException {
        return processRequest("createCampaignEmail", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap createCampaignFAX(EgoiMap arguments) throws EgoiException {
        return processRequest("createCampaignFAX", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap createCampaignSMS(EgoiMap arguments) throws EgoiException {
        return processRequest("createCampaignSMS", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap createCampaignVoice(EgoiMap arguments) throws EgoiException {
        return processRequest("createCampaignVoice", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap createList(EgoiMap arguments) throws EgoiException {
        return processRequest("createList", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap createSegment(EgoiMap arguments) throws EgoiException {
        return processRequest("createSegment", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap deleteCampaign(EgoiMap arguments) throws EgoiException {
        return processRequest("deleteCampaign", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap deleteList(EgoiMap arguments) throws EgoiException {
        return processRequest("deleteList", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap deleteSegment(EgoiMap arguments) throws EgoiException {
        return processRequest("deleteSegment", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap editCampaignEmail(EgoiMap arguments) throws EgoiException {
        return processRequest("editCampaignEmail", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap editCampaignSMS(EgoiMap arguments) throws EgoiException {
        return processRequest("editCampaignSMS", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap editExtraField(EgoiMap arguments) throws EgoiException {
        return processRequest("editExtraField", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap editSubscriber(EgoiMap arguments) throws EgoiException {
        return processRequest("editSubscriber", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMapList getCampaigns(EgoiMap arguments) throws EgoiException {
        return processRequest("getCampaigns", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMap getClientData(EgoiMap arguments) throws EgoiException {
        return processRequest("getClientData", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap getCredits(EgoiMap arguments) throws EgoiException {
        return processRequest("getCredits", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap getReport(EgoiMap arguments) throws EgoiException {
        return processRequest("getReport", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMapList getSegments(EgoiMap arguments) throws EgoiException {
        return processRequest("getSegments", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMapList getSenders(EgoiMap arguments) throws EgoiException {
        return processRequest("getSenders", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMap removeSubscriber(EgoiMap arguments) throws EgoiException {
        return processRequest("removeSubscriber", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap sendCall(EgoiMap arguments) throws EgoiException {
        return processRequest("sendCall", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap sendEmail(EgoiMap arguments) throws EgoiException {
        return processRequest("sendEmail", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap sendFAX(EgoiMap arguments) throws EgoiException {
        return processRequest("sendFAX", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap sendSMS(EgoiMap arguments) throws EgoiException {
        return processRequest("sendSMS", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap subscriberData(EgoiMap arguments) throws EgoiException {
        return processRequest("subscriberData", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap updateList(EgoiMap arguments) throws EgoiException {
        return processRequest("updateList", arguments, EgoiMap.class);
    }

    //NEW STUFF
    @Override
    public EgoiMapList getAutobots(EgoiMap arguments) throws EgoiException {
        return processRequest("getAutobots", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMap startAutobot(EgoiMap arguments) throws EgoiException {
        return processRequest("startAutobot", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap createCampaignGroup(EgoiMap arguments) throws EgoiException {
        return processRequest("createCampaignGroup", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap editCampaignGroup(EgoiMap arguments) throws EgoiException {
        return processRequest("editCampaignGroup", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMapList getHeaderFooterTemplates(EgoiMap arguments) throws EgoiException {
        return processRequest("getHeaderFooterTemplates", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMapList getTemplates(EgoiMap arguments) throws EgoiException {
        return processRequest("getTemplates", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMap editApiCallback(EgoiMap arguments) throws EgoiException {
        return processRequest("editApiCallback", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMapList getOperationsQueue(EgoiMap arguments) throws EgoiException {
        return processRequest("getOperationsQueue", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMapList getForms(EgoiMap arguments) throws EgoiException {
        return processRequest("getForms", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMapList getFormsSubmissions(EgoiMap arguments) throws EgoiException {
        return processRequest("getFormsSubmissions", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMap deleteExtraField(EgoiMap arguments) throws EgoiException {
        return processRequest("deleteExtraField", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMapList getExtraFields(EgoiMap arguments) throws EgoiException {
        return processRequest("getExtraFields", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMap generateAdvancedReport(EgoiMap arguments) throws EgoiException {
        return processRequest("generateAdvancedReport", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMapList getAdvancedReport(EgoiMap arguments) throws EgoiException {
        return processRequest("getAdvancedReport", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMap createSegmentAdvanced(EgoiMap arguments) throws EgoiException {
        return processRequest("createSegmentAdvanced", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap addCname(EgoiMap arguments) throws EgoiException {
        return processRequest("addCname", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap addSender(EgoiMap arguments) throws EgoiException {
        return processRequest("addSender", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap deleteSender(EgoiMap arguments) throws EgoiException {
        return processRequest("deleteSender", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap editSender(EgoiMap arguments) throws EgoiException {
        return processRequest("editSender", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMapList getSubscriberActivity(EgoiMap arguments) throws EgoiException {
        return processRequest("getSubscriberActivity", arguments, EgoiMapList.class);
    }

    @Override
    public EgoiMap addTag(EgoiMap arguments) throws EgoiException {
        return processRequest("addTag", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap attachTag(EgoiMap arguments) throws EgoiException {
        return processRequest("attachTag", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap deleteTag(EgoiMap arguments) throws EgoiException {
        return processRequest("deleteTag", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMap detachTag(EgoiMap arguments) throws EgoiException {
        return processRequest("detachTag", arguments, EgoiMap.class);
    }

    @Override
    public EgoiMapList getTags(EgoiMap arguments) throws EgoiException {
        return processRequest("getTags", arguments, EgoiMapList.class);
    }

}

File: e-goi-wrapper-csharp/EgoiApiProject/EgoiApi/Impl/XmlRpc/EgoiApiXmlRpcImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CookComputing.XmlRpc;

namespace Egoi
{
    public class EgoiApiXmlRpcImpl : EgoiApi
    {

        private static EgoiApi instance;

        private EgoiApiXmlRpcProxy proxy;

        private EgoiApiXmlRpcImpl() {
            proxy = XmlRpcProxyGen.Create<EgoiApiXmlRpcProxy>();
        }

        public static EgoiApi getInstance()
        {
            if (instance == null)
                instance = new EgoiApiXmlRpcImpl();
            return instance;
        }

        private XmlRpcStruct[] encodeMapList(EgoiMapList list)
        {
            List<XmlRpcStruct> slist = new List<XmlRpcStruct>();

            foreach (EgoiMap map in list)
                slist.Add(encodeMap(map));

            return slist.ToArray<XmlRpcStruct>();
        }

        private XmlRpcStruct encodeMap(EgoiMap map)
        {
            XmlRpcStruct ret = new XmlRpcStruct();
            foreach (String k in map.Keys)
            {
                object value = map[k];
                if (value is string)
                    ret[k] = value;

                else if (value is EgoiMap)
                    ret[k] = encodeMap(value as EgoiMap);

                else if (value is EgoiMapList)
                    ret[k] = encodeMapList(value as EgoiMapList);
                else
                    // convert to string
                    ret[k] = value.ToString();


                // ignore other types

            }
            return ret;
        }

        private void checkResult(object result)
        {
            if (result == null)
                throw new EgoiException("The response from the server was null");

            if (result is string)
                throw new EgoiException(EgoiException.idToMessage(result as string));
        }

        private EgoiMap decodeMap(XmlRpcStruct result)
        {
            EgoiMap map = new EgoiMap();
            foreach (object _k in result.Keys)
            {
                if (_k is string)
                {
                    string k = _k as string;
                    object value = result[_k];
                    if (value is string)
                        map[k] = value as string;

                    if (value is XmlRpcStruct)
                        map[k] = decodeMap(value as XmlRpcStruct);

                    if (value is XmlRpcStruct[])
                        map[k] = decodeMapList(value as XmlRpcStruct[]);

                    if(value is int)
                        map[k] = value as int?;

                    if (value is double)
                        map[k] = value as double?;

                    if (value is float)
                        map[k] = value as float?;

                }
            }

            return map;
        }

        private EgoiMap decodeResultMap(object result)
        {
            checkResult(result);

            if (result is XmlRpcStruct)
                return decodeMap(result as XmlRpcStruct);
            else
                throw new EgoiException("The response is of unexpected type: " + result + ", expecting XmlRpcStruct");
        }

        private EgoiMapList decodeMapList(XmlRpcStruct[] result)
        {
            EgoiMapList list = new EgoiMapList();
            foreach (XmlRpcStruct value in result)
                list.Add(decodeMap(value));
            return list;
        }

        private EgoiMapList decodeResultMapList(object result)
        {
            checkResult(result);

            if (result is XmlRpcStruct[])
                return decodeMapList(result as XmlRpcStruct[]);
            else if (result is object[]) // occurs when response is empty
                return new EgoiMapList();
            else
                throw new EgoiException("The response is of unexpected type: " + result + ", expecting XmlRpcStruct[]");
        }

        ////////////////////////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////////////////////////

        public EgoiMap addExtraField(EgoiMap arguments)
        {
            return decodeResultMap(proxy.addExtraField(encodeMap(arguments)));
        }

        public EgoiMap addSubscriber(EgoiMap arguments)
        {
            return decodeResultMap(proxy.addSubscriber(encodeMap(arguments)));
        }

        public EgoiMap addSubscriberBulk(EgoiMap arguments)
        {
            return decodeResultMap(proxy.addSubscriberBulk(encodeMap(arguments)));
        }

        public EgoiMap checklogin(EgoiMap arguments)
        {
            return decodeResultMap(proxy.checklogin(encodeMap(arguments)));
        }

        public EgoiMap createCampaignEmail(EgoiMap arguments)
        {
            return decodeResultMap(proxy.createCampaignEmail(encodeMap(arguments)));
        }

        public EgoiMap createCampaignFAX(EgoiMap arguments)
        {
            return decodeResultMap(proxy.createCampaignFAX(encodeMap(arguments)));
        }

        public EgoiMap createCampaignSMS(EgoiMap arguments)
        {
            return decodeResultMap(proxy.createCampaignSMS(encodeMap(arguments)));
        }

        public EgoiMap createCampaignVoice(EgoiMap arguments)
        {
            return decodeResultMap(proxy.createCampaignVoice(encodeMap(arguments)));
        }

        public EgoiMap createList(EgoiMap arguments)
        {
            return decodeResultMap(proxy.createList(encodeMap(arguments)));
        }

        public EgoiMap createSegment(EgoiMap arguments)
        {
            return decodeResultMap(proxy.createSegment(encodeMap(arguments)));
        }

        public EgoiMap deleteCampaign(EgoiMap arguments)
        {
            return decodeResultMap(proxy.deleteCampaign(encodeMap(arguments)));
        }

        public EgoiMap deleteSegment(EgoiMap arguments)
        {
            return decodeResultMap(proxy.deleteSegment(encodeMap(arguments)));
        }

        public EgoiMap editCampaignEmail(EgoiMap arguments)
        {
            return decodeResultMap(proxy.editCampaignEmail(encodeMap(arguments)));
        }

        public EgoiMap editCampaignSMS(EgoiMap arguments)
        {
            return decodeResultMap(proxy.editCampaignSMS(encodeMap(arguments)));
        }

        public EgoiMap editExtraField(EgoiMap arguments)
        {
            return decodeResultMap(proxy.editExtraField(encodeMap(arguments)));
        }

        public EgoiMap editSubscriber(EgoiMap arguments)
        {
            return decodeResultMap(proxy.editSubscriber(encodeMap(arguments)));
        }

        public EgoiMapList getCampaigns(EgoiMap arguments)
        {
            return decodeResultMapList(proxy.getCampaigns(encodeMap(arguments)));
        }

        public EgoiMap getClientData(EgoiMap arguments)
        {
            return decodeResultMap(proxy.getClientData(encodeMap(arguments)));
        }

        public EgoiMap getCredits(EgoiMap arguments)
        {
            return decodeResultMap(proxy.getCredits(encodeMap(arguments)));
        }

        public EgoiMapList getLists(EgoiMap arguments)
        {
            return decodeResultMapList(proxy.getLists(encodeMap(arguments)));
        }

        public EgoiMap getReport(EgoiMap arguments)
        {
            return decodeResultMap(proxy.getReport(encodeMap(arguments)));
        }

        public EgoiMapList getSegments(EgoiMap arguments)
        {
            return decodeResultMapList(proxy.getSegments(encodeMap(arguments)));
        }

        public EgoiMapList getSenders(EgoiMap arguments)
        {
            return decodeResultMapList(proxy.getSenders(encodeMap(arguments)));
        }

        public EgoiMap getUserData(EgoiMap arguments)
        {
            return decodeResultMap(proxy.getUserData(encodeMap(arguments)));
        }

        public EgoiMap removeSubscriber(EgoiMap arguments)
        {
            return decodeResultMap(proxy.removeSubscriber(encodeMap(arguments)));
        }

        public EgoiMap sendCall(EgoiMap arguments)
        {
            return decodeResultMap(proxy.sendCall(encodeMap(arguments)));
        }

        public EgoiMap sendEmail(EgoiMap arguments)
        {
            return decodeResultMap(proxy.sendEmail(encodeMap(arguments)));
        }

        public EgoiMap sendFAX(EgoiMap arguments)
        {
            return decodeResultMap(proxy.sendFAX(encodeMap(arguments)));
        }

        public EgoiMap sendSMS(EgoiMap arguments)
        {
            return decodeResultMap(proxy.sendSMS(encodeMap(arguments)));
        }

        public EgoiMap subscriberData(EgoiMap arguments)
        {
            return decodeResultMap(proxy.subscriberData(encodeMap(arguments)));
        }

        public EgoiMap updateList(EgoiMap arguments)
        {
            return decodeResultMap(proxy.updateList(encodeMap(arguments)));
        }

    }
}
